File: 0020_valid_parentheses/is_valid.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

// Time: O(n)
// Space: O(n)
bool is_valid(string s)
{
    stack<char> stack;
    for (auto c : s)
    {
        switch (c)
        {
        case '{':
            stack.push('}');
            break;
        case '[':
            stack.push(']');
            break;
        case '(':
            stack.push(')');
            break;
        default:
            if (!stack.size() || stack.top() != c)
                return false;
            stack.pop();
        }
    }
    return stack.empty();
}

File: 0020_valid_parentheses/is_valid.py
def is_valid(s: str) -> bool:
    """
    Time: O(n)
    Space: O(n)
    """
    map = {
        "}": "{",
        "]": "[",
        ")": "("
    }
    stack = []
    for c in s:
        if c in map:
            if not stack or stack.pop() != map[c]:
                return False
        else:
            stack.append(c)
    return len(stack) == 0
